### PR TITLE
Fix: DO-2607 investigate propose solution tiers time series graph in f cose layout is squashed

### PR DESCRIPTION
**Pull request by patricia-causalens**
## Motivation and Context

For more complex time series graphs where not all nodes had a time_lag aspect the graph could end up looking squished.

image

## Implementation Description

Instead of using `variable_name` to set the layers, we instead look through the nodes and for those that share a `variable_name` we set a new property which is then used to add only those to layers. The other nodes are free to be distributed as the layout algorithm normally would.

## Any new dependencies Introduced

None

## How Has This Been Tested?

Locally on storybook and with the graph where the original problem was found.

## PR Checklist:

- [x] I have implemented all requirements? (see JIRA, project documentation).
- [x] I am not affecting someone else's work, If I am, they are included as a reviewer.
- [ ] I have added relevant tests (unit, integration or regression).
- [ ] I have added comments to all the bits that are hard to follow.
- [ ] I have added/updated Documentation.
- [x] I have updated the appropriate changelog with a line for my changes.

## Screenshots (if appropriate):

Updated graph after changes:
Screenshot 2024-03-05 at 13 49 07

### CODE REVIEW

File: packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx [hunk 1]
```diff
@@ -63,35 +96,35 @@ export default function useCausalGraphEditor(
...
+
         if (
             isTimeSeriesCausalGraph &&
             'tiers' in newLayout &&
             'orientation' in newLayout &&
             newLayout.tiers === undefined
         ) {
-            // We set tiers based on TimeSeriesCausalGraph structure of node lags
-            newLayout.tiers = { group: 'variable_name', order_nodes_by: 'time_lag' };
+            updateNodesForTimeSeries(state.graph, api);
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx:123]: It's an anti-pattern to fire side effects within a `useMemo`, can we think of a better way to do it?

File: packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx [hunk 2]
```diff
@@ -108,7 +141,7 @@ export default function useCausalGraphEditor(
 
         lastParentData.current = graphData;
         // eslint-disable-next-line react-hooks/exhaustive-deps
-    }, [graphData]);
+    }, []);
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx:144]: Why are we removing the dependency here? Without it it will always only compare to the original graph data

**Comment by tamara-causalens** [comment]
Thanks Patricia! This looks quite far away from the graph the user expects to see (given that the original graph was meant to be Planar). Have we considered defaulting to a planar layout without tiers (or set to none) when planar with tiers fails instead of forcing the layout to be fcose?

**Comment by patricia-causalens** [comment]
> Thanks Patricia! This looks quite far away from the graph the user expects to see (given that the original graph was meant to be Planar). Have we considered defaulting to a planar layout without tiers (or set to none) when planar with tiers fails instead of forcing the layout to be fcose?

Based on the ticket title I fixed Fcose so that it wouldn't be squashed, this would also fix it for other layouts. This is for improving TimeSeries layouts in general.

If now we go to the Planar case which does not currently support time series, we can add a guard so that if it is a TimeSeriesCausalGraph and PlanarLayout has been set that we will not apply tiers unless the user has set these explicitly.

File: packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx (hunk 2 above)
Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx]: oops it should not have been, added it back

File: packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx (hunk 1 above)
Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx]: After reading the useEffect page you suggested, from what I understood their recommended approach is to be outside of a memo.

File: packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx [hunk 3]
```diff
@@ -93,6 +106,23 @@ export default function useCausalGraphEditor(
...
+
+    // If the graph is a time series graph, tiers are not defined and is not a PlanarLayout, we update the tiers to show the time series in layers
+    if (
+        isTimeSeriesCausalGraph &&
+        'tiers' in graphLayout &&
+        'orientation' in graphLayout &&
+        graphLayout.tiers === undefined &&
+        !(graphLayout instanceof PlanarLayout)
+    ) {
+        updateNodesForTimeSeries(state.graph, api);
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx:122]: The difference between this code and the canonical example in https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes is that here we're mutating state within a class and not invoking `setState` anywhere so this won't have the same behaviour of restarting a new render.

Though I'm not too worried about the graph itself, since we're not really tying it to React UI deliberately - the engine is isolated so it's a bit of a special case.

What I would be worried about is that:
1) you're mutating a prop here (in the form of `graphLayout`) - see https://react.dev/learn/keeping-components-pure ; it would be fine if you created a modified clone of the layout and return it from here
2) this will rerun on every render it seems like? again in the example from effect docs this pattern works because within the if clause they are calling `setState` which makes the next re-render *not* go inside the if again, preventing circular renders 

I feel like we're conflating two things here:
1) updating the graphLayout based on some conditions -> this could be a useMemo and return a new graph layout here
2) adjusting the graph state to contain tiers based on some conditions -> this feels like it belongs in the parser or should be run together with the parser before updating the state - so in 1) the reducer initializer, and 2) in the effect which invokes INIT_GRAPH before

I understand the condition might be the same for both (and we can extract it to a helper or something) but these are fundamentally different things - 1) is an adjustment to the graph state which should be done where the state is put into the reducer state, and 2) is an adjustment of graph layout which is not stored anywhere and just returned, so it can be constructed and returned (and wrapped in a useMemo but that's an optimization)

I hope this makes sense, let me know and we can chat about it more. 

Also cc @Roman-Kornev, feel free to chime in if you have some thoughts on how to not turn it into more of a spaghetti than it already is (also my fault) 😅

Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx:122]: Ok I see what you mean, will give it another go with the parser, that does seem like it might be a good place for it to be added

Comment (Roman-Kornev) [reply packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx:122]: Agreed, `graphLayout` is just derived state so it should go into a `useMemo`.

The trouble with `updateNodesForTimeSeries` also lies in the fact that it only updates in 1 direction: from non-timeseries to timeseries. Can we convert the graph from timeseries back to non-timeseries? In that case those extras would stay, but should they? All those checks should probably go in the parser, yeah.

Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx]: It is not updating a non time series to time series graph, it is just adding another property if it is a time series graph so that we can use internally to assign the layers.

File: packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx [hunk 4]
```diff
@@ -63,36 +64,36 @@ export default function useCausalGraphEditor(
...
             isTimeSeriesCausalGraph &&
             'tiers' in newLayout &&
             'orientation' in newLayout &&
-            newLayout.tiers === undefined
+            newLayout.tiers === undefined &&
+            !(newLayout instanceof PlanarLayout)
         ) {
-            // We set tiers based on TimeSeriesCausalGraph structure of node lags
-            newLayout.tiers = { group: 'variable_name', order_nodes_by: 'time_lag' };
+            newLayout.tiers = { group: 'extras.time_series_variable', order_nodes_by: 'time_lag' };
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx:92]: I think we missed it before but `newLayout` is a reference to the same layout object as `graphLayout` which ends up mutating the prop. I think best would be to deep clone the layout when creating the copy.

Since layout is a class instance it will be slightly tricky to clone, but a lodash clone (or `structuredClone`) and maybe also explicitly setting the prototype of the clone should we enough.

File: packages/ui-causal-graph-editor/src/shared/parsers.tsx [hunk 5]
```diff
@@ -303,6 +334,11 @@ export function causalGraphParser(
         }
     });
 
+    // If the graph is a time series graph, we add a property to the nodes that share the same variable_name so that they can be later grouped into layers if needed
+    if (Object.values(data.nodes)[0]?.node_class === 'TimeSeriesNode') {
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/src/shared/parsers.tsx:338]: Can we cover this in a parser test?

**Review by krzysztof-causalens — COMMENTED** [review]
Looks much better, left a couple more comments

File: packages/ui-causal-graph-editor/src/shared/parsers.tsx (hunk 5 above)
Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/shared/parsers.tsx:338]: Added

File: packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx (hunk 4 above)
Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx:92]: structuredClone and lodash's didn't quite work because they would loose the notion of the instance of the layout class. So instead I added a clone method to GraphLayout

Comment (krzysztof-causalens) [reply packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx:92]: Yeah they don't work by themselves, that's why I suggested you might also need to set the prototype. I wonder if this works, e.g.
```ts
Object.create(
    Object.getPrototypeOf(obj), 
    Object.getOwnPropertyDescriptors(obj) 
)
```
or something in that vein. 
Regardless, your approach is more 'proper' though a bit more verbose and we'll just need to remember to keep the clone method up to date as we add more properties

Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx:96]: That does work, have updated to use it instead

**Review by krzysztof-causalens — APPROVED** [review]
